### PR TITLE
[WIP] Meta-set nodedef: Part 4

### PR DESCRIPTION
**Pull request by Thomas--S**
See #5895 for part 1
See #5902 for part 2
See #5903 for part 3

Part 1-3 are the base for actual meta-set nodedef.

Real meta-set nodedef arrives here in part 4.

Until now the following is possible:

- [ ] `on_construct`
- [ ] `on_destruct`
- [ ] `after_destruct`
- [ ] `on_rightclick`
- [ ] `groups`
- [ ] `drawtype`
- [ ] `visual_scale`
- [ ] `mesh`
- [ ] `tiles`
- [ ] `overlay_tiles`
- [ ] `special_tiles`
- [ ] `alpha`
- [ ] `use_texture_alpha`
- [ ] `color`
- [ ] `palette`
- [x] **`post_effect_color`**
- [ ] `paramtype`
- [ ] `paramtype2`
- [ ] `is_ground_content` _(possibly senseless)_
- [ ] `sunlight_propagates`
- [ ] `walkable`
- [x] **`pointable`**
- [ ] `diggable`
- [x] **`climbable`**
- [x] **`buildable_to`**
- [ ] `floodable`
- [ ] `liquidtype`
- [ ] `liquid_alternative_flowing`
- [ ] `liquid_alternative_source`
- [ ] `liquid_viscosity`
- [ ] `liquid_range`
- [ ] `leveled`
- [ ] `liquidrenewable`
- [x] **`drowning`**
- [ ] `light_source`
- [x] **`damage_per_second`**
- [ ] `node_box`
- [ ] `connects_to`
- [ ] `connect_sides`
- [x] **`selection_box`**
- [x] **`collision_box`**
- [ ] `waving`
- [ ] `sounds`

### CODE REVIEW

**Comment by Thomas--S** [comment]
Example for how to set the collision box with meta-set nodedef:

```lua
minetest.register_on_punchnode(function(pos, node, puncher, pointed_thing)
	minetest.set_def(pos, {
		collision_box = {
			type = "fixed",
			fixed = {
				{ -0.15, -0.5, -0.15, 0.15, 0.5, 0.15 },
			},
		}
	})
end)
```

**Comment by Thomas--S** [comment]
I decided not to consider meta-set nodedef during mapgen. I think this is not so important, and if necessary it can be added later.

**Comment by bigfoot547** [comment]
What happened to this PR? Have you stopped working on it?

**Comment by Thomas--S** [comment]
At the moment I'm quite busy with school. I plan to work on it again in a few weeks.

**Comment by bigfoot547** [comment]
Okay, thanks @Thomas--S

**Comment by sofar** [comment]
The most important param I see that's missing is `tiles`.

Take care of school, let me know when you're back available to work on this and I'd like to test.

**Comment by Thomas--S** [comment]
The holidays started two weeks ago, but I haven't worked on this yet. I plan to continue with meta-set nodedef soon.

Some time ago, I already tried to make tiles work, but there were some serious problem. I think I'll try again to implement tiles.

I'd be very glad if you could review part 1 in the meantime, as it is already ready. This will help to avoid too many rebases.

File: src/mapblock.cpp [hunk 1]
```diff
@@ -108,6 +109,72 @@ MapBlock::~MapBlock()
...
+
+HybridPtr<const ContentFeatures> MapBlock::getNodeDefNoCheck(v3s16 p, bool *valid_position)
+{
+	*valid_position = isValidPosition(p);
+	if(!valid_position) {
+		return NULL;
+	}
+	// If special definition exists, return a shared pointer to a copy of it
+	if(m_special_nodedefs.count(p))
+		return new ContentFeatures(m_special_nodedefs[p]);
```
Comment (numberZero) [inline src/mapblock.cpp]: @Thomas--S May the pointer outlive the `MapBlock`? If not, why not return `m_special_nodedefs[p]` itself?

**Comment by Thomas--S** [comment]
I didn't find time for this during the holidays, I won't be able to finish this before June 2018. I'm also thinking of a different approach. So, I suggest not to merge any of my meta-set nodedef PR's yet.
So, I'll also close part 1-3, which are all contained in part 4.

**Comment by paramat** [comment]
Does this include celeron55's suggestion?
https://github.com/minetest/minetest/pull/1118#issuecomment-75022588

> I think it could be a good idea to store the difference in node definition as a serialized Lua table in the node metadata instead of the actual serialized C++ node definition or parts of it. This would radically improve the cross-version compatibility, which is a serious problem in the serialized node definition (it's not designed for this, and should not need to be designed for this), and would allow easily storing only the fields that have been modified. It would also be more understandable to modders.

What about kwolekr's concerns?
https://github.com/minetest/minetest/pull/1118#issuecomment-116349795

> By the way, what does HybridPtr do and why is it necessary for this code?
I am strongly opposed to using smart/refcounting/scoped pointer template classes as this makes object lifetime much more difficult to track and adds semantics that may be unfamiliar and counterproductive with some developers.
The claim that it code is 'safer' using smart pointer is a dubious one at best. We should strive to not fall for flavor-of-the-month coding "best practices".

Has this speed issue been fixed?
https://github.com/minetest/minetest/pull/1118#issuecomment-120691098

> I did not work on it lately, but last time I tried, there were bugs (changing a node def could take up to 2s (!!)), and I wasn't able to get rid of that speed problem.

> If someone could fix the speed problem, it would be the most important thing to do.

Also has this been fixed?
https://github.com/minetest/minetest/pull/1118#issuecomment-220854685

> This should be fixed, but that part of the code wasn't finished anyway.

Have all comments and concerns in #1118 been addressed?

**Comment by paramat** [comment]
> I won't be able to finish this before June 2018. I'm also thinking of a different approach.

Closing for now but to be reopened when progress continues.

**Comment by paramat** [comment]
See https://github.com/minetest/minetest/pull/7278#issuecomment-384629820
PR is abandoned.

**Pull request closed without merging**